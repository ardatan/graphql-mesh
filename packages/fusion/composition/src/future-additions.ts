import type { YamlConfig } from '@graphql-mesh/types';

export type FutureAddition = YamlConfig.AdditionalStitchingResolverObject;
export const futureAdditions: FutureAddition[] = [];
