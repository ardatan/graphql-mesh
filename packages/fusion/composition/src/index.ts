export * from './compose.js';
export * from './transforms.js';
export * from './federationSupergraph.js';
