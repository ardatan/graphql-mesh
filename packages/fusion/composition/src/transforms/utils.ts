export class TransformValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'TransformValidationError';
  }
}
