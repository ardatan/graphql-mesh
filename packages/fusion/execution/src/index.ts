export * from './execution.js';
export * from './extractSubgraph.js';
export * from './flattenSelections.js';
export * from './operations.js';
export * from './query-planning.js';
export * from './serialization.js';
export * from './types.js';
