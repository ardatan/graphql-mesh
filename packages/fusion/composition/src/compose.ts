import {
  getNamedType,
  GraphQLField,
  GraphQLNamedType,
  GraphQLSchema,
  isObjectType,
  isSpecifiedScalarType,
  OperationTypeNode,
} from 'graphql';
import { pascalCase } from 'pascal-case';
import { snakeCase } from 'snake-case';
import { mergeSchemas, MergeSchemasConfig } from '@graphql-tools/schema';
import { asArray, getRootTypeMap, getRootTypes, MapperKind, mapSchema } from '@graphql-tools/utils';
import { getDirectiveExtensions } from './getDirectiveExtensions.js';

export interface SubgraphConfig {
  name: string;
  schema: GraphQLSchema;
  transforms?: SubgraphTransform[];
}

export type SubgraphTransform = (
  schema: GraphQLSchema,
  subgraphConfig: SubgraphConfig,
) => GraphQLSchema;

const defaultRootTypeNames: Record<OperationTypeNode, string> = {
  query: 'Query',
  mutation: 'Mutation',
  subscription: 'Subscription',
};

type ResolverAnnotation = {
  subgraph: string;
  operation: string;
  kind: 'BATCH' | 'FETCH';
};

type VariableAnnotation = {
  subgraph: string;
  name: string;
  select: string;
};

export function composeSubgraphs(
  subgraphs: SubgraphConfig[],
  options?: Omit<MergeSchemasConfig, 'schema'>,
) {
  const annotatedSubgraphs: GraphQLSchema[] = [];
  const transformedSubgraphMap = new Map<string, GraphQLSchema>();
  for (const subgraphConfig of subgraphs) {
    const { name: subgraphName, schema, transforms } = subgraphConfig;
    const rootTypeMap = getRootTypeMap(schema);
    const typeToOperationType = new Map<string, OperationTypeNode>();

    for (const [operationType, rootType] of rootTypeMap) {
      typeToOperationType.set(rootType.name, operationType);
    }

    const queryType = schema.getQueryType();
    const queryFields = queryType?.getFields();

    const annotatedSubgraph = mapSchema(schema, {
      [MapperKind.TYPE]: type => {
        if (isSpecifiedScalarType(type)) {
          return type;
        }
        const operationType = typeToOperationType.get(type.name);
        if (operationType) {
          return new (Object.getPrototypeOf(type).constructor)({
            ...type.toConfig(),
            name: defaultRootTypeNames[operationType],
          });
        }
        const directives: Record<string, any> = {
          ...getDirectiveExtensions(type),
          source: {
            subgraph: subgraphName,
            name: type.name,
          },
        };
        // Automatic type merging configuration based on ById and ByIds naming conventions before transforms
        addAnnotationsForSemanticConventions({
          type,
          queryFields,
          subgraphName,
          directives,
          subgraphs,
          transformedSubgraphMap,
        });
        return new (Object.getPrototypeOf(type).constructor)({
          ...type.toConfig(),
          extensions: {
            ...type.extensions,
            directives,
          },
        });
      },
      [MapperKind.FIELD]: (fieldConfig, fieldName) => ({
        ...fieldConfig,
        extensions: {
          ...fieldConfig.extensions,
          directives: {
            ...getDirectiveExtensions(fieldConfig),
            source: {
              subgraph: subgraphName,
              name: fieldName,
              type: fieldConfig.type.toString(),
            },
          },
        },
      }),
      [MapperKind.ENUM_VALUE]: (valueConfig, _typeName, _schema, externalValue) => ({
        ...valueConfig,
        extensions: {
          ...valueConfig.extensions,
          directives: {
            ...getDirectiveExtensions(valueConfig),
            source: {
              subgraph: subgraphName,
              name: externalValue,
            },
          },
        },
      }),
      [MapperKind.ROOT_FIELD]: (fieldConfig, fieldName, typeName) => {
        const operationType = typeToOperationType.get(typeName);
        const operationName =
          operationType === 'query' ? fieldName : `${operationType}${fieldName}`;
        const variableDefinitions: string[] = [];
        const rootFieldArgs: string[] = [];
        if (fieldConfig.args) {
          for (const argName in fieldConfig.args) {
            const arg = fieldConfig.args[argName];
            let variableDefinitionStr = `$${argName}: ${arg.type}`;
            if (arg.defaultValue) {
              variableDefinitionStr += ` = ${
                typeof arg.defaultValue === 'string'
                  ? JSON.stringify(arg.defaultValue)
                  : arg.defaultValue
              }`;
            }
            variableDefinitions.push(variableDefinitionStr);
            rootFieldArgs.push(`${argName}: $${argName}`);
          }
        }
        const variableDefinitionsString = variableDefinitions.length
          ? `(${variableDefinitions.join(', ')})`
          : '';
        const rootFieldArgsString = rootFieldArgs.length ? `(${rootFieldArgs.join(', ')})` : '';
        const operationString = `${operationType} ${operationName}${variableDefinitionsString} { ${fieldName}${rootFieldArgsString} }`;

        return {
          ...fieldConfig,
          extensions: {
            ...fieldConfig.extensions,
            directives: {
              ...getDirectiveExtensions(fieldConfig),
              resolver: {
                subgraph: subgraphName,
                operation: operationString,
              },
              source: {
                subgraph: subgraphName,
                name: fieldName,
                type: fieldConfig.type.toString(),
              },
            },
          },
        };
      },
    });

    let transformedSubgraph = annotatedSubgraph;
    transformedSubgraphMap.set(subgraphName, transformedSubgraph);
    if (transforms?.length) {
      for (const transform of transforms) {
        transformedSubgraph = transform(transformedSubgraph, subgraphConfig);
        transformedSubgraphMap.set(subgraphName, transformedSubgraph);
      }
      // Semantic conventions
      const transformedQueryType = transformedSubgraph.getQueryType();
      const transformedQueryFields = transformedQueryType?.getFields();
      const rootTypes = getRootTypes(transformedSubgraph);
      transformedSubgraph = mapSchema(transformedSubgraph, {
        [MapperKind.TYPE]: type => {
          if (isSpecifiedScalarType(type) || rootTypes.has(type as any)) {
            return type;
          }
          const directives = getDirectiveExtensions(type);
          // Automatic type merging configuration based on ById and ByIds naming conventions after transforms
          addAnnotationsForSemanticConventions({
            type,
            queryFields: transformedQueryFields,
            subgraphName,
            directives,
            subgraphs,
            transformedSubgraphMap,
          });
          return new (Object.getPrototypeOf(type).constructor)({
            ...type.toConfig(),
            extensions: {
              ...type.extensions,
              directives,
            },
          });
        },
      });
    }
    annotatedSubgraphs.push(transformedSubgraph);
    transformedSubgraphMap.set(subgraphName, transformedSubgraph);
  }

  return mergeSchemas({
    schemas: annotatedSubgraphs,
    assumeValidSDL: true,
    assumeValid: true,
    ...options,
  });
}

function addAnnotationsForSemanticConventions({
  type,
  queryFields,
  subgraphName,
  directives,
  subgraphs,
  transformedSubgraphMap,
}: {
  type: GraphQLNamedType;
  queryFields?: Record<string, GraphQLField<any, any>>;
  subgraphName: string;
  directives: Record<string, any>;
  subgraphs: SubgraphConfig[];
  transformedSubgraphMap: Map<string, GraphQLSchema>;
}) {
  if (queryFields && isObjectType(type)) {
    const fieldMap = type.getFields();
    for (const queryFieldName in queryFields) {
      for (const fieldName in fieldMap) {
        const objectField = fieldMap[fieldName];
        const queryField = queryFields[queryFieldName];
        const arg = queryField.args.find(
          arg => getNamedType(arg.type) === getNamedType(objectField.type),
        );
        const queryFieldTypeName = getNamedType(queryField.type).name;
        const queryFieldNameSnakeCase = snakeCase(queryFieldName);
        const varName = `${type.name}_${fieldName}`;
        if (arg && queryFieldTypeName === type.name) {
          // eslint-disable-next-line no-inner-declarations
          function addVariablesForOtherSubgraphs() {
            directives.variable ||= [];
            for (const otherSubgraphConfig of subgraphs) {
              const otherType = otherSubgraphConfig.schema.getType(type.name);
              const otherTransformedType = transformedSubgraphMap
                .get(otherSubgraphConfig.name)
                ?.getType(type.name);
              const otherTypeFieldNames = [];
              if (isObjectType(otherType)) {
                otherTypeFieldNames.push(...Object.keys(otherType.getFields()));
              }
              if (isObjectType(otherTransformedType)) {
                otherTypeFieldNames.push(...Object.keys(otherTransformedType.getFields()));
              }
              if (otherTypeFieldNames.includes(fieldName)) {
                directives.variable ||= [];
                if (
                  !directives.variable.some(
                    (v: VariableAnnotation) =>
                      v.subgraph === otherSubgraphConfig.name && v.name === varName,
                  )
                ) {
                  directives.variable.push({
                    subgraph: otherSubgraphConfig.name,
                    name: varName,
                    select: fieldName,
                  });
                }
              }
            }
          }
          switch (queryFieldNameSnakeCase) {
            case snakeCase(type.name):
            case snakeCase(`get_${type.name}_by_${fieldName}`):
            case snakeCase(`${type.name}_by_${fieldName}`): {
              const operationName = pascalCase(`${type.name}_by_${fieldName}`);
              const originalFieldName = getOriginalFieldNameForSubgraph(queryField, subgraphName);
              const resolverAnnotation: ResolverAnnotation = {
                subgraph: subgraphName,
                operation: `query ${operationName}($${varName}: ${arg.type}) { ${originalFieldName}(${arg.name}: $${varName}) }`,
                kind: 'FETCH',
              };
              directives.resolver ||= [];
              directives.resolver.push(resolverAnnotation);
              addVariablesForOtherSubgraphs();
              break;
            }
            case snakeCase(`${type.name}s`):
            case snakeCase(`get_${type.name}s_by_${fieldName}`):
            case snakeCase(`${type.name}s_by_${fieldName}`):
            case snakeCase(`get_${type.name}s_by_${fieldName}s`):
            case snakeCase(`${type.name}s_by_${fieldName}s`): {
              const operationName = pascalCase(`${type.name}s_by_${fieldName}s`);
              const originalFieldName =
                getOriginalFieldNameForSubgraph(queryField, subgraphName) || queryFieldName;
              const resolverAnnotation: ResolverAnnotation = {
                subgraph: subgraphName,
                operation: `query ${operationName}($${varName}: ${arg.type}) { ${originalFieldName}(${arg.name}: $${varName}) }`,
                kind: 'BATCH',
              };
              directives.resolver ||= [];
              directives.resolver.push(resolverAnnotation);
              directives.variable ||= [];
              addVariablesForOtherSubgraphs();
              break;
            }
          }
        }
      }
    }
  }
}

function getOriginalFieldNameForSubgraph(
  field: GraphQLField<any, any>,
  subgraph: string,
): string | void {
  if (field.extensions?.directives) {
    const sourceDirectives = asArray((field.extensions.directives as any).source);
    const sourceDirective = sourceDirectives.find((d: any) => d.subgraph === subgraph);
    if (sourceDirective) {
      return sourceDirective.name;
    }
  }
  return field.name;
}
