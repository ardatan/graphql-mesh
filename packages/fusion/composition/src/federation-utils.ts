import { GraphQLDirective, isSpecifiedDirective, type GraphQLSchema } from 'graphql';
import { getDirectiveExtensions, MapperKind, mapSchema } from '@graphql-tools/utils';

export function addFederation2DirectivesToSubgraph(subgraph: GraphQLSchema) {
  const schemaDirectives = getDirectiveExtensions(subgraph);
  const linkDirectives = (schemaDirectives.link ||= []);
  if (
    !linkDirectives.some(linkDirectiveArgs =>
      linkDirectiveArgs.url?.startsWith('https://specs.apollo.dev/link/'),
    )
  ) {
    linkDirectives.push({
      url: 'https://specs.apollo.dev/link/v1.0',
    });
  }
  if (
    !linkDirectives.some(linkDirectiveArgs =>
      linkDirectiveArgs.url?.startsWith('https://specs.apollo.dev/federation/'),
    )
  ) {
    linkDirectives.push({
      url: 'https://specs.apollo.dev/federation/v2.3',
      import: [],
    });
  }
  const extensions: Record<string, unknown> = (subgraph.extensions ||= {});
  extensions.directives = schemaDirectives;
  return subgraph;
}

export function importFederationDirectives(subgraph: GraphQLSchema, directives: string[]) {
  const schemaDirectives = getDirectiveExtensions(subgraph);
  const linkDirectives = (schemaDirectives.link ||= []);
  let importStatement = linkDirectives.find(linkDirectiveArgs =>
    linkDirectiveArgs.url?.startsWith('https://specs.apollo.dev/federation/'),
  );
  if (!importStatement) {
    importStatement = {
      url: 'https://specs.apollo.dev/federation/v2.3',
      import: [],
    };
    linkDirectives.push(importStatement);
  }
  importStatement.import = [...new Set([...(importStatement.import || []), ...directives])];
  const extensions: Record<string, unknown> = (subgraph.extensions ||= {});
  extensions.directives = schemaDirectives;
  return subgraph;
}

export function importMeshDirectives(subgraph: GraphQLSchema, directives: string[]) {
  const schemaDirectives = getDirectiveExtensions(subgraph) || {};
  const linkDirectives = (schemaDirectives.link ||= []);
  let importStatement = linkDirectives.find(linkDirectiveArgs =>
    linkDirectiveArgs.url?.startsWith('https://the-guild.dev/graphql/mesh/spec/'),
  );
  if (!importStatement) {
    importStatement = {
      url: 'https://the-guild.dev/graphql/mesh/spec/v1.0',
      import: [],
    };
    linkDirectives.push(importStatement);
  }
  importStatement.import = [...new Set([...importStatement.import, ...directives])];
  const composeDirectives = (schemaDirectives.composeDirective = schemaDirectives.composeDirective
    ? schemaDirectives.composeDirective.filter(Boolean)
    : []);
  for (const directiveName of directives) {
    if (!composeDirectives.some(dir => dir.name === directiveName)) {
      composeDirectives.push({
        name: directiveName,
      });
    }
  }
  const extensions: Record<string, unknown> = (subgraph.extensions ||= {});
  extensions.directives = schemaDirectives;
  subgraph = importFederationDirectives(subgraph, ['@composeDirective']);
  return subgraph;
}

const FEDERATION_V1_DIRECTIVES = [
  '@key',
  '@provides',
  '@requires',
  '@external',
  '@inaccessible',
  '@shareable',
  '@extends',
  '@tag',
];

export function detectAndAddMeshDirectives(subgraph: GraphQLSchema) {
  const meshDirectives: string[] = [];
  subgraph = mapSchema(subgraph, {
    [MapperKind.DIRECTIVE]: directive => {
      const directiveName = `@${directive.name}`;
      if (!isSpecifiedDirective(directive) && !FEDERATION_V1_DIRECTIVES.includes(directiveName)) {
        meshDirectives.push(directiveName);
        if (!directive.isRepeatable && directive.args.some(arg => arg.name === 'subgraph')) {
          return new GraphQLDirective({
            ...directive.toConfig(),
            isRepeatable: true,
          });
        }
      }
      return directive;
    },
  });
  if (meshDirectives.length > 0) {
    subgraph = importMeshDirectives(subgraph, meshDirectives);
  }
  return subgraph;
}

export function convertSubgraphToFederationv2(subgraph: GraphQLSchema) {
  const schemaDirectives = getDirectiveExtensions(subgraph);
  const linkDirectives = (schemaDirectives.link ||= []);
  if (
    linkDirectives.some(linkDirectiveArgs =>
      linkDirectiveArgs.url?.startsWith('https://specs.apollo.dev/link/'),
    )
  ) {
    // This is already v2, skipping
    return subgraph;
  }
  subgraph = addFederation2DirectivesToSubgraph(subgraph);
  // Add @shareable
  subgraph = mapSchema(subgraph, {
    [MapperKind.OBJECT_TYPE]: type => {
      const typeDirectives = getDirectiveExtensions(type);
      typeDirectives.shareable ||= [];
      if (!typeDirectives.shareable?.length) {
        const typeExtensions: Record<string, any> = (type.extensions ||= {});
        typeExtensions.directives ||= {};
        typeExtensions.directives.shareable = [{}];
      }
      return type;
    },
    [MapperKind.INTERFACE_TYPE]: type => {
      if ((type?.extensions?.directives as any)?.shareable) {
        delete (type.extensions.directives as any).shareable;
      }
      return type;
    },
  });
  subgraph = importFederationDirectives(subgraph, FEDERATION_V1_DIRECTIVES);
  return subgraph;
}
