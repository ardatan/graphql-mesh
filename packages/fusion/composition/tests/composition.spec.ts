import { buildSchema } from 'graphql';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { composeSubgraphs, createRenameFieldTransform, createRenameTypeTransform } from '../src';

describe('Composition', () => {
  const aSchema = buildSchema(/* GraphQL */ `
    type Query {
      myFoo: Foo!
    }

    type Foo {
      id: ID!
    }
  `);

  const bSchema = buildSchema(/* GraphQL */ `
    type Query {
      foo(id: ID!): Foo!
      foos(ids: [ID!]!): [Foo!]!
    }

    type Foo {
      id: ID!
      bar: String!
    }
  `);
  it('composes basic schemas', () => {
    const composedSchema = composeSubgraphs([
      {
        name: 'A',
        schema: aSchema,
      },
      {
        name: 'B',
        schema: bSchema,
      },
    ]);

    expect(printSchemaWithDirectives(composedSchema)).toMatchSnapshot();
  });
  it('composes with transforms on types', () => {
    const prefixTransform = createRenameTypeTransform(
      (type, subgraphConfig) => `${subgraphConfig.name}_${type.name}`,
    );
    const composedSchema = composeSubgraphs([
      {
        name: 'A',
        schema: aSchema,
        transforms: [prefixTransform],
      },
      {
        name: 'B',
        schema: bSchema,
        transforms: [prefixTransform],
      },
    ]);

    expect(printSchemaWithDirectives(composedSchema)).toMatchSnapshot();
  });
  it('composes with transforms on fields', () => {
    const prefixTransform = createRenameFieldTransform(
      (_field, fieldName, _typeName, subgraphConfig) => `${subgraphConfig.name}_${fieldName}`,
    );
    const composedSchema = composeSubgraphs([
      {
        name: 'A',
        schema: aSchema,
        transforms: [prefixTransform],
      },
      {
        name: 'B',
        schema: bSchema,
        transforms: [prefixTransform],
      },
    ]);

    expect(printSchemaWithDirectives(composedSchema)).toMatchSnapshot();
  });
  it('keeps the directives', () => {
    const composedSchema = composeSubgraphs([
      {
        name: 'A',
        schema: aSchema,
      },
      {
        name: 'B',
        schema: bSchema,
      },
      {
        name: 'C',
        schema: buildSchema(/* GraphQL */ `
          directive @foo on FIELD_DEFINITION
          type Query {
            cFoo: Foo! @foo
          }

          type Foo {
            id: ID!
          }
        `),
      },
    ]);

    expect(printSchemaWithDirectives(composedSchema)).toMatchSnapshot();
  });
});
