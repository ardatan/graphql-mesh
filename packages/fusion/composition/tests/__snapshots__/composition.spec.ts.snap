// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Composition composes basic schemas 1`] = `
"schema {
  query: Query
}

type Query {
  myFoo: Foo! @resolver(subgraph: "A", operation: "query myFoo { myFoo }") @source(subgraph: "A", name: "myFoo", type: "Foo!")
  foo(id: ID!): Foo! @resolver(subgraph: "B", operation: "query foo($id: ID!) { foo(id: $id) }") @source(subgraph: "B", name: "foo", type: "Foo!")
  foos(ids: [ID!]!): [Foo!]! @resolver(subgraph: "B", operation: "query foos($ids: [ID!]!) { foos(ids: $ids) }") @source(subgraph: "B", name: "foos", type: "[Foo!]!")
}

type Foo @source(subgraph: "A", name: "Foo") @source(subgraph: "B", name: "Foo") @resolver(subgraph: "B", operation: "query FooById($Foo_id: ID!) { foo(id: $Foo_id) }", kind: "FETCH") @resolver(subgraph: "B", operation: "query FoosByIds($Foo_id: [ID!]!) { foos(ids: $Foo_id) }", kind: "BATCH") @variable(subgraph: "A", name: "Foo_id", select: "id") @variable(subgraph: "B", name: "Foo_id", select: "id") {
  id: ID! @source(subgraph: "A", name: "id", type: "ID!") @source(subgraph: "B", name: "id", type: "ID!")
  bar: String! @source(subgraph: "B", name: "bar", type: "String!")
}"
`;

exports[`Composition composes with transforms on fields 1`] = `
"schema {
  query: Query
}

type Query {
  A_myFoo: Foo! @resolver(subgraph: "A", operation: "query myFoo { myFoo }") @source(subgraph: "A", name: "myFoo", type: "Foo!")
  B_foo(id: ID!): Foo! @resolver(subgraph: "B", operation: "query foo($id: ID!) { foo(id: $id) }") @source(subgraph: "B", name: "foo", type: "Foo!")
  B_foos(ids: [ID!]!): [Foo!]! @resolver(subgraph: "B", operation: "query foos($ids: [ID!]!) { foos(ids: $ids) }") @source(subgraph: "B", name: "foos", type: "[Foo!]!")
}

type Foo @source(subgraph: "A", name: "Foo") @source(subgraph: "B", name: "Foo") @resolver(subgraph: "B", operation: "query FooById($Foo_id: ID!) { foo(id: $Foo_id) }", kind: "FETCH") @resolver(subgraph: "B", operation: "query FoosByIds($Foo_id: [ID!]!) { foos(ids: $Foo_id) }", kind: "BATCH") @variable(subgraph: "A", name: "Foo_id", select: "id") @variable(subgraph: "B", name: "Foo_id", select: "id") {
  A_id: ID! @source(subgraph: "A", name: "id", type: "ID!")
  B_id: ID! @source(subgraph: "B", name: "id", type: "ID!")
  B_bar: String! @source(subgraph: "B", name: "bar", type: "String!")
}"
`;

exports[`Composition composes with transforms on types 1`] = `
"schema {
  query: B_Query
}

type A_Query {
  myFoo: A_Foo! @resolver(subgraph: "A", operation: "query myFoo { myFoo }") @source(subgraph: "A", name: "myFoo", type: "Foo!")
}

type A_Foo @source(subgraph: "A", name: "Foo") {
  id: ID! @source(subgraph: "A", name: "id", type: "ID!")
}

type B_Query {
  foo(id: ID!): B_Foo! @resolver(subgraph: "B", operation: "query foo($id: ID!) { foo(id: $id) }") @source(subgraph: "B", name: "foo", type: "Foo!")
  foos(ids: [ID!]!): [B_Foo!]! @resolver(subgraph: "B", operation: "query foos($ids: [ID!]!) { foos(ids: $ids) }") @source(subgraph: "B", name: "foos", type: "[Foo!]!")
}

type B_Foo @source(subgraph: "B", name: "Foo") @resolver(subgraph: "B", operation: "query FooById($Foo_id: ID!) { foo(id: $Foo_id) }", kind: "FETCH") @resolver(subgraph: "B", operation: "query FoosByIds($Foo_id: [ID!]!) { foos(ids: $Foo_id) }", kind: "BATCH") @variable(subgraph: "A", name: "Foo_id", select: "id") @variable(subgraph: "B", name: "Foo_id", select: "id") {
  id: ID! @source(subgraph: "B", name: "id", type: "ID!")
  bar: String! @source(subgraph: "B", name: "bar", type: "String!")
}"
`;

exports[`Composition keeps the directives 1`] = `
"schema {
  query: Query
}

directive @foo on FIELD_DEFINITION

type Query {
  myFoo: Foo! @resolver(subgraph: "A", operation: "query myFoo { myFoo }") @source(subgraph: "A", name: "myFoo", type: "Foo!")
  foo(id: ID!): Foo! @resolver(subgraph: "B", operation: "query foo($id: ID!) { foo(id: $id) }") @source(subgraph: "B", name: "foo", type: "Foo!")
  foos(ids: [ID!]!): [Foo!]! @resolver(subgraph: "B", operation: "query foos($ids: [ID!]!) { foos(ids: $ids) }") @source(subgraph: "B", name: "foos", type: "[Foo!]!")
  cFoo: Foo! @foo @resolver(subgraph: "C", operation: "query cFoo { cFoo }") @source(subgraph: "C", name: "cFoo", type: "Foo!")
}

type Foo @source(subgraph: "A", name: "Foo") @source(subgraph: "B", name: "Foo") @source(subgraph: "C", name: "Foo") @resolver(subgraph: "B", operation: "query FooById($Foo_id: ID!) { foo(id: $Foo_id) }", kind: "FETCH") @resolver(subgraph: "B", operation: "query FoosByIds($Foo_id: [ID!]!) { foos(ids: $Foo_id) }", kind: "BATCH") @variable(subgraph: "A", name: "Foo_id", select: "id") @variable(subgraph: "B", name: "Foo_id", select: "id") @variable(subgraph: "C", name: "Foo_id", select: "id") {
  id: ID! @source(subgraph: "A", name: "id", type: "ID!") @source(subgraph: "B", name: "id", type: "ID!") @source(subgraph: "C", name: "id", type: "ID!")
  bar: String! @source(subgraph: "B", name: "bar", type: "String!")
}"
`;
